reactive-governance/cas-reactive/apply-labels/automate_project_labels.py

```diff
@@ -0,0 +1,257 @@
+
+
+def build_labels_dict(df):
+    """
+    Build dictionary object with project id as key and dictionary of labels as value
+        Parameters:
+            df (data frame): CSV data converted to data frame
+        Returns:
+            all project labels (dict): returns dict with project_id as key and
+            nested dict of labels as key:value pair for each project_id
+    """
+    all_projects_labels = {}
+    columns = df.columns
+    for index, row in df.iterrows():
+        project_labels = {}
+        for column in columns:
+            # skip project columns and cells where value is empty
+            if column != "project_id" and not pd.isnull(row[column]):
+                project_labels[column] = row[column]
+        json_str = json.dumps(project_labels, indent=4)
+        print(f"project_id: {row[columns[0]]}, \nlabels: {json_str}\n")
+        all_projects_labels[row[columns[0]]] = project_labels
+    return all_projects_labels
+
+
+def apply_labels(all_project_labels):
+    """
+    Apply labels for all project_ids extracted from the csv
+        Parameters:
+            all_project_labels (nested dictionary): dictionary contains project_ids
+            and corresponding labels
+    """
+    for project_id in all_project_labels.keys():
+        labels = all_project_labels[project_id]
+        # print(f"key: {project_id}, \nvalue: {labels}\n")
+        apply_project_labels(project_id, labels)
+
+
+def parse_csv(df):
+    columns = df.columns
+    print("\n=== Validating CSV ===")
+    print("========================")
+    key_pattern = r'^([a-z])([a-z0-9_-]{0,62})$'
+    value_pattern = r'^([a-z0-9_-]{0,62})?$'
+    warning_count = 0
+    error_count = 0
+
+    for key in columns:
+        if "Unnamed" in key:
+            warning_count += 1
+            logging.warning("Found empty column in the csv")
+            print("WARNING: Found empty column in the csv")
+        if not re.match(key_pattern, key):
+            error_count += 1
+            logging.error(f"Label Key '{key}' does not follow recommended pattern.")
+            print(f"ERROR: Label Key '{key}' does not follow recommended pattern. "
+                  f"Refer https://cloud.google.com/resource-manager/docs/labels-overview#requirements")
+    for index, row in df.iterrows():
+        for column in columns:
+            if not pd.isnull(row[column]) and not re.match(value_pattern, row[column]):
+                error_count += 1
+                logging.error(f" Label Value '{row[column]}' does not follow recommended pattern.")
+                print(f"ERROR: Label Value '{row[column]}' does not follow recommended pattern. "
+                      f"Refer https://cloud.google.com/resource-manager/docs/labels-overview#requirements")
+    return error_count
+
+
+def exit_program():
+    print("Exiting the program...")
+    sys.exit(0)
+
+
+# ENTER CSV file url from GCS bucket
+CSV_GS_LOCATION = "gs://cost_attribution/labels_update.csv"
```

Comment:
Doesn't make sense make it more dynamic? Like: 

- A parameter when calling the python
- An environment variable present in VM created through terraform

What do you think?

Reply:
Thank you. Updated code to provide CSV file path as a parameter.